src/utils/span32.cairo

```diff
@@ -0,0 +1,176 @@
+) -> Option<Array<T>> {
+    if remaining == 0 {
+        return Option::Some(curr_output);
+    }
+    curr_output.append(TSerde::deserialize(ref serialized)?);
+    deserialize_array_helper(ref serialized, curr_output, remaining - 1)
+}
+
+impl SpanCopy<T> of Copy<Span32<T>>;
+impl SpanDrop<T> of Drop<Span32<T>>;
+
+impl Span32Serde<T, impl TSerde: Serde<T>, impl TDrop: Drop<T>> of Serde<Span32<T>> {
+    fn serialize(self: @Span32<T>, ref output: Array<felt252>) {
+        (*self).len().serialize(ref output);
+        serialize_array_helper(*self, ref output)
+    }
+
+    fn deserialize(ref serialized: Span<felt252>) -> Option<Span32<T>> {
+        let length = *serialized.pop_front()?;
+        let mut arr: Array<T> = ArrayTrait::new();
+        Option::Some(Array32Trait::span(@deserialize_array_helper(ref serialized, arr, length)?))
+    }
+}
+
+#[generate_trait]
+impl Span32Impl<T, impl TSerde: Serde<T>> of Span32Trait<T> {
+    #[inline(always)]
+    fn pop_front(ref self: Span32<T>) -> Option<@T> {
+        self.snapshot.pop_front()
+    }
+    #[inline(always)]
+    fn pop_back(ref self: Span32<T>) -> Option<@T> {
+        self.snapshot.pop_back()
+    }
+    #[inline(always)]
+    fn get(self: Span32<T>, index: usize) -> Option<Box<@T>> {
+        self.snapshot.get(index)
+    }
+    #[inline(always)]
+    fn at(self: Span32<T>, index: usize) -> @T {
+        self.snapshot.at(index)
+    }
+    #[inline(always)]
+    fn slice(self: Span32<T>, start: usize, length: usize) -> Span32<T> {
+        Span32 { snapshot: self.snapshot.slice(start, length) }
+    }
+    #[inline(always)]
+    fn len(self: Span32<T>) -> usize {
+        self.snapshot.len()
+    }
+    #[inline(always)]
+    fn is_empty(self: Span32<T>) -> bool {
+        self.snapshot.is_empty()
+    }
+}
+
+impl Span32Index<T> of IndexView<Span32<T>, usize, @T> {
+    #[inline(always)]
+    fn index(self: @Span32<T>, index: usize) -> @T {
+        self.snapshot.index(index)
+    }
+}
+
+trait Array32Trait<T> {
+    fn span(self: @Array<T>) -> Span32<T>;
+}
+
+impl Array32<T> of Array32Trait<T> {
+    #[inline(always)]
+    fn span(self: @Array<T>) -> Span32<T> {
+        assert(self.len() <= 32, 'array too big');
+        Span32 { snapshot: Span { snapshot: self } }
+    }
+}
+
+impl StoreContractAddressSpan of Store<Span32<ContractAddress>> {
+    fn read(
+        address_domain: u32, base: StorageBaseAddress
+    ) -> SyscallResult<Span32<ContractAddress>> {
+        StoreContractAddressSpan::read_at_offset(address_domain, base, 0)
+    }
+
+    fn write(
+        address_domain: u32, base: StorageBaseAddress, value: Span32<ContractAddress>
+    ) -> SyscallResult<()> {
+        StoreContractAddressSpan::write_at_offset(address_domain, base, 0, value)
+    }
+
+    fn read_at_offset(
+        address_domain: u32, base: StorageBaseAddress, mut offset: u8
+    ) -> SyscallResult<Span32<ContractAddress>> {
+        let mut arr: Array<ContractAddress> = ArrayTrait::new();
```

Comment:
is it maybe easier to call `Store::<Span>::read_at_offset` and wrap the returned Span inside the Span32 struct?